### ISSUE TITLE
ERRORDATA_STACK_SIZE exceeded error thrown during some of the calculations

### ISSUE THREAD

**Issue by sokolval** [greenplum-db/gpdb#17192]
### Greenplum version or build

 PostgreSQL 9.4.26 (Greenplum Database 6.26.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609, 64-bit compiled on Jan 10 2024 14:21:32

### OS version and uname -a

Ubuntu Xenial Linux 5.4.196-37 https://github.com/greenplum-db/gpdb/pull/1 SMP Thu May 26 12:08:09 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### autoconf options used ( config.status --config )

./configure --with-perl --with-python --without-gssapi --disable-gpcloud --with-libxml --with-system-tzdata=/usr/share/zoneinfo --prefix=/usr/local/gpdb LDFLAGS=-L/usr/lib/jni/ --with-pam --with-openssl

### Installation information ( pg_config )

BINDIR = /usr/local/gpdb/bin
DOCDIR = /usr/local/gpdb/share/doc/postgresql
HTMLDIR = /usr/local/gpdb/share/doc/postgresql
INCLUDEDIR = /usr/local/gpdb/include
PKGINCLUDEDIR = /usr/local/gpdb/include/postgresql
INCLUDEDIR-SERVER = /usr/local/gpdb/include/postgresql/server
LIBDIR = /usr/local/gpdb/lib/x86_64-linux-gnu
PKGLIBDIR = /usr/local/gpdb/lib/x86_64-linux-gnu/postgresql
LOCALEDIR = /usr/local/gpdb/share/locale
MANDIR = /usr/local/gpdb/share/man
SHAREDIR = /usr/local/gpdb/share/postgresql
SYSCONFDIR = /etc/postgresql
PGXS = /usr/local/gpdb/lib/x86_64-linux-gnu/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--build=x86_64-linux-gnu' '--prefix=/usr' '--includedir=/usr/local/gpdb/include' '--mandir=/usr/local/gpdb/share/man' '--infodir=/usr/local/gpdb/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=/usr/local/gpdb/lib/x86_64-linux-gnu' '--libexecdir=/usr/local/gpdb/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--with-perl' '--with-python' '--without-gssapi' '--disable-gpcloud' '--with-libxml' '--with-system-tzdata=/usr/share/zoneinfo' '--prefix=/usr/local/gpdb' 'LDFLAGS=-L/usr/lib/jni/' '--with-pam' '--with-openssl' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security'
CC = gcc
CPPFLAGS = -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2 -I/usr/local/gpdb/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/usr/local/gpdb/include
CFLAGS_SL = -fPIC
LDFLAGS = -L/usr/lib/jni/ -Wl,--as-needed -Wl,-rpath,'/usr/local/gpdb/lib/x86_64-linux-gnu',--enable-new-dtags -L/usr/local/gpdb/lib/x86_64-linux-gnu
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -lxerces-c -lxml2 -lpam -lrt -lzstd -lrt -lcrypt -ldl -lm -L/usr/local/gpdb/lib/x86_64-linux-gnu
VERSION = PostgreSQL 9.4.26

### Expected behavior

Calculations work normally and errors are thrown in understandable format, without overflowing errordata stack

### Actual behavior

Some queries run longer than expected and cause ERRORDATA_STACK_SIZE exceeded error while normally they should run in few seconds or less. e.g. temp table is created:
```
CREATE TEMPORARY TABLE "step_1" ON COMMIT DROP AS
SELECT "_e_59359dc236_sk" as "e_59359dc236_e1__id", "_e_b519f5f7ca_sk" as "e_b519f5f7ca_e2__id"
    , "_e_5d03a163e9_sk" as "e_5d03a163e9_e3__id", coalesce("deleted_flg", false) as "_deleted_flg"
    , _business_date AS utc_valid_from_dttm
    , _etl_processed_dttm
    
FROM "result_e_59359dc236_new_be330f7a3e"
WHERE (("_e_59359dc236_sk" IS NOT NULL and "_e_b519f5f7ca_sk" IS NOT NULL)) AND ((True )) AND (TRUE AND ("_e_5d03a163e9_sk" IS NOT NULL = "_e_5d03a163e9_sk_exist"))
DISTRIBUTED BY ("e_59359dc236_e1__id")
```
This above table is created in less than second and contains few (<100) rows.
After that command is called:
```
CREATE TEMPORARY TABLE "distinct_keys" ON COMMIT DROP AS
SELECT DISTINCT "e_59359dc236_e1__id", "e_b519f5f7ca_e2__id" FROM "step_1"
DISTRIBUTED BY ("e_59359dc236_e1__id")
```
This command hangs for more than 1 hour, and after that throws ERRORDATA_STACK_SIZE exceeded error on all segments simultaneously which is totally unexpected

### Step to reproduce the behavior

Not clear, I will try to reproduce with example data, but generally it is not repeating, I mean on the second run all works perfect.

I checked dmesg and syslog - no errors there
Resgroup is not congested at the moment. The Vmem consumption is showed as 10 to 13 GB per cluster per this query which seems to be not so much.

1 notifiable statement is that on resource queue this type of error never thrown.

It seems to be somehow related to [Issue #16531](https://github.com/greenplum-db/gpdb/pull/16531) but i do not have clear evidence for that.

**Comment by kainwen** [comment]
Do you have coredump that can be shared here?
Or can you please use gdb to read the error data in the core?

**Comment by sokolval** [comment]
I tried examining process nearly to break with perf utility, however it did not give much details, all looks standard, except for the wile looking at main MPPEXEC select process I received below messages:
```
         ? (         ):  ... [continued]: select())                                           = 0 (Timeout)
                                       select (/lib/x86_64-linux-gnu/libc-2.23.so)
                                       format_sockaddr (/usr/local/gpdb/bin/postgres)
     0.020 (500.339 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
                                       select (/lib/x86_64-linux-gnu/libc-2.23.so)
                                       format_sockaddr (/usr/local/gpdb/bin/postgres)
   500.377 (500.582 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
                                       select (/lib/x86_64-linux-gnu/libc-2.23.so)
                                       format_sockaddr (/usr/local/gpdb/bin/postgres)
  1000.978 (500.376 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
                                       select (/lib/x86_64-linux-gnu/libc-2.23.so)
                                       format_sockaddr (/usr/local/gpdb/bin/postgres)
  1501.371 (500.587 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
                                       select (/lib/x86_64-linux-gnu/libc-2.23.so)
                                       format_sockaddr (/usr/local/gpdb/bin/postgres)
  2001.979 (500.340 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
                                       select (/lib/x86_64-linux-gnu/libc-2.23.so)
                                       format_sockaddr (/usr/local/gpdb/bin/postgres)
```

In terms of using gdb - could you recomment any specific arguments to run it and investigate efficiently?

**Comment by kainwen** [comment]
> I tried examining process nearly to break with perf utility, however it did not give much details, all looks standard, except for the wile looking at main MPPEXEC select process I received below messages:
> 
> ```
>          ? (         ):  ... [continued]: select())                                           = 0 (Timeout)
>                                        select (/lib/x86_64-linux-gnu/libc-2.23.so)
>                                        format_sockaddr (/usr/local/gpdb/bin/postgres)
>      0.020 (500.339 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
>                                        select (/lib/x86_64-linux-gnu/libc-2.23.so)
>                                        format_sockaddr (/usr/local/gpdb/bin/postgres)
>    500.377 (500.582 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
>                                        select (/lib/x86_64-linux-gnu/libc-2.23.so)
>                                        format_sockaddr (/usr/local/gpdb/bin/postgres)
>   1000.978 (500.376 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
>                                        select (/lib/x86_64-linux-gnu/libc-2.23.so)
>                                        format_sockaddr (/usr/local/gpdb/bin/postgres)
>   1501.371 (500.587 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
>                                        select (/lib/x86_64-linux-gnu/libc-2.23.so)
>                                        format_sockaddr (/usr/local/gpdb/bin/postgres)
>   2001.979 (500.340 ms): select(n: 202, inp: 0x7ffd4c6bd530, tvp: 0x7ffd4c6bd520)              = 0 (Timeout)
>                                        select (/lib/x86_64-linux-gnu/libc-2.23.so)
>                                        format_sockaddr (/usr/local/gpdb/bin/postgres)
> ```
> 
> In terms of using gdb - could you recomment any specific arguments to run it and investigate efficiently?

@sokolval 

`ERRORDATA_STACK_SIZE` should PANIC and generate core dumps in your OS if you enabled it.
core dump path is recorded in the kernel parameter: `/proc/sys/kernel/core_pattern`, you can read the path by `cat /proc/sys/kernel/core_pattern` and then go to the directory to see if there is core dump.

Besides in the errored segment log, you should find LOGs containing call stacks, this can be used to double confirm the core (with pid in the LOG).

Now suppose you have got the coredump:

1. source `greenplum_path.sh` in your OS
2. `which postgres`, this command will tell you the full path of `postgres` binary
3. `gdb <full path of postgres binary> -c <core_dump_file_path>` this is use gdb to look into the core dump
4. when reach here, you should be inside the gdb:
    - `bt` to show the backtrace
    - ` p errordata[0]` to show the top element in error stack
    - ` p errordata[1]` and ` p errordata[2]` , ..... ` p errordata[6]` to show some more

with all the above information (LOGS and GDB coredump analysis output) we might better understand the problem.

If you core file is share-able, you can use `packcore` to pack it and send it to us with GPDB LOGs.

**Comment by sokolval** [comment]
this is an output from some of the gdb commands:
(gdb) info stack
#0  0x00007f035e72c6b3 in select () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000000000b5fdee in ?? ()
#2  0x0000000000b632ac in TeardownTCPInterconnect ()
#3  0x0000000000b5dad9 in TeardownInterconnect ()
#4  0x00000000008923ff in mppExecutorFinishup ()
#5  0x000000000087f848 in standard_ExecutorEnd ()
#6  0x0000000000830a88 in PortalCleanup ()
#7  0x0000000000b270ea in PortalDrop ()
#8  0x00000000009f22a7 in ?? ()
#9  0x00000000009f57b8 in PostgresMain ()
#10 0x00000000006b583e in ?? ()
#11 0x0000000000990309 in PostmasterMain ()
#12 0x00000000006b839b in main ()
(gdb) info program
	Using the running image of attached Thread 0x7f0361606740 (LWP 678676).
Program stopped at 0x7f035e72c6b3.
Type "info stack" or "info registers" for more information.
(gdb) info program
	Using the running image of attached Thread 0x7f0361606740 (LWP 678676).
Program stopped at 0x7f035e72c6b3.
Type "info stack" or "info registers" for more information.
(gdb) info program
	Using the running image of attached Thread 0x7f0361606740 (LWP 678676).
Program stopped at 0x7f035e72c6b3.
Type "info stack" or "info registers" for more information.
(gdb) info stack
#0  0x00007f035e72c6b3 in select () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000000000b5fdee in ?? ()
#2  0x0000000000b632ac in TeardownTCPInterconnect ()
#3  0x0000000000b5dad9 in TeardownInterconnect ()
#4  0x00000000008923ff in mppExecutorFinishup ()
#5  0x000000000087f848 in standard_ExecutorEnd ()
#6  0x0000000000830a88 in PortalCleanup ()
#7  0x0000000000b270ea in PortalDrop ()
#8  0x00000000009f22a7 in ?? ()
#9  0x00000000009f57b8 in PostgresMain ()
#10 0x00000000006b583e in ?? ()
#11 0x0000000000990309 in PostmasterMain ()
#12 0x00000000006b839b in main ()

**Comment by kainwen** [comment]
> this is an output from some of the gdb commands:
> 
> ```
> (gdb) info stack
> #0  0x00007f035e72c6b3 in select () from /lib/x86_64-linux-gnu/libc.so.6
> #1  0x0000000000b5fdee in ?? ()
> #2  0x0000000000b632ac in TeardownTCPInterconnect ()
> #3  0x0000000000b5dad9 in TeardownInterconnect ()
> #4  0x00000000008923ff in mppExecutorFinishup ()
> #5  0x000000000087f848 in standard_ExecutorEnd ()
> #6  0x0000000000830a88 in PortalCleanup ()
> #7  0x0000000000b270ea in PortalDrop ()
> #8  0x00000000009f22a7 in ?? ()
> #9  0x00000000009f57b8 in PostgresMain ()
> #10 0x00000000006b583e in ?? ()
> #11 0x0000000000990309 in PostmasterMain ()
> #12 0x00000000006b839b in main ()
> 
> (gdb) info program
> 	Using the running image of attached Thread 0x7f0361606740 (LWP 678676).
> Program stopped at 0x7f035e72c6b3.
> Type "info stack" or "info registers" for more information.
> 
> (gdb) info program
> 	Using the running image of attached Thread 0x7f0361606740 (LWP 678676).
> Program stopped at 0x7f035e72c6b3.
> Type "info stack" or "info registers" for more information.
> 
> (gdb) info program
> 	Using the running image of attached Thread 0x7f0361606740 (LWP 678676).
> Program stopped at 0x7f035e72c6b3.
> Type "info stack" or "info registers" for more information.
> 
> (gdb) info stack
> #0  0x00007f035e72c6b3 in select () from /lib/x86_64-linux-gnu/libc.so.6
> #1  0x0000000000b5fdee in ?? ()
> #2  0x0000000000b632ac in TeardownTCPInterconnect ()
> #3  0x0000000000b5dad9 in TeardownInterconnect ()
> #4  0x00000000008923ff in mppExecutorFinishup ()
> #5  0x000000000087f848 in standard_ExecutorEnd ()
> #6  0x0000000000830a88 in PortalCleanup ()
> #7  0x0000000000b270ea in PortalDrop ()
> #8  0x00000000009f22a7 in ?? ()
> #9  0x00000000009f57b8 in PostgresMain ()
> #10 0x00000000006b583e in ?? ()
> #11 0x0000000000990309 in PostmasterMain ()
> #12 0x00000000006b839b in main ()
> 
> (gdb) info registers
> rax            0xfffffffffffffdfe	-514
> rbx            0x4	4
> rcx            0x7f035e72c6b3	139652446209715
> rdx            0x0	0
> rsi            0x7fff9565d320	140735699866400
> rdi            0x8b	139
> rbp            0x1	0x1
> rsp            0x7fff9565b2d8	0x7fff9565b2d8
> r8             0x7fff9565b300	140735699858176
> r9             0x7fff9565b2b0	140735699858096
> r10            0x0	0
> r11            0x246	582
> r12            0x3480cf8	55053560
> r13            0x3480cf0	55053552
> r14            0x80	128
> r15            0x8a	138
> rip            0x7f035e72c6b3	0x7f035e72c6b3 <select+19>
> eflags         0x246	[ PF ZF IF ]
> cs             0x33	51
> ss             0x2b	43
> ds             0x0	0
> es             0x0	0
> fs             0x0	0
> gs             0x0	0
> k0             0x0	0
> k1             0x0	0
> k2             0x0	0
> k3             0x0	0
> k4             0x0	0
> k5             0x0	0
> k6             0x0	0
> k7             0x0	0
> 
> (gdb) info frame
> Stack level 0, frame at 0x7fff9565b2e0:
>  rip = 0x7f035e72c6b3 in select; saved rip = 0xb5fdee
>  called by frame at 0x7fff9565f370
>  Arglist at 0x7fff9565b2d0, args: 
>  Locals at 0x7fff9565b2d0, Previous frame's sp is 0x7fff9565b2e0
>  Saved registers:
>   rip at 0x7fff9565b2d8
> ```

Can you print `errordata` as I sugguested in preivous comments?

**Comment by sokolval** [comment]
I tried printing errordata as well as errordata[0] and arrordata[1], etc
However the result is as follows:
```
(gdb) p errordata
No symbol table is loaded.  Use the "file" command.
(gdb) p errordata[0]
No symbol table is loaded.  Use the "file" command.
```

Could you advise if I need to apply some flag when configure or build gpdb to make symbols included?
Another possibility - maybe I can get a Symbols file from somewhere to supply it to gdb separately (with --symbols=SYMFILE  Read symbols from SYMFILE.)?

In fact we tried to b uild gpdb with corefiles enabled, and it caused gpdb to produce coredump and restart cluster on any non-significant error, which was not acceptable even for TEST environment..

**Comment by sokolval** [comment]
I was able to build GP postgres executable with symbols and now waiting for suitable process to examine

**Comment by interma** [comment]
Hi @sokolval 
Based on your provided logs, one of them is notable:
```
2024-03-15 12:37:38.207942,"gpadmin","test",p1028215,th-1930782912,"master_ip","16150",2024-03-15 12:37:19,3670769,con165168,cmd990,seg19,slice3,dx1191684,x3670769,sx1,"LOG","58M01","Interconnect timeout: Connection to seg78 [seg_n_IP]:41127 from local port was not complete after 4000ms 4020 e
lapsed. Will retry.",,,,,,"CREATE TEMPORARY TABLE ""tmp_windows_for_join""
```
The related code is in `SetupTCPInterconnect()` and `setupOutgoingConnection()`, the logic is very simple: 
Just try a tcp connect.

So, `Interconnect timeout: Connection to seg78 [seg_n_IP]:41127 from local port was not complete after 4000/8000/12000/...ms` means master cannot tcp connect to seg78.

Is it possible to have some network failure in your env? And when it reproduces, you can immediately do a manual try (e.g. using `nc` command) between the host pair to see if can establish a tcp connection.

**Comment by sokolval** [comment]
Thank you, I will catch the situation and try to check connection either with nc or telnet. Will get back to you once I have logs/facts!

**Comment by sokolval** [comment]
Hello!
We've investigated on our side and found out that we faced an overflow in net connections table which is regulated by kernel parameter **net.netfilter.nf_conntrack_max**. Since we have heavy load on GP, having 100+ segments, it seems that this parameter capacity was not enough to withhold all the demand for IP connections.
Hence currently we raised thus param to 5898240 and errors were gone. 
When we had it set to 2097152 it was not enough. 
Under heavy load this error may leave traces in syslog similar to:  **w13 kernel: [1012462.328789] nf_conntrack: table full, dropping packet**

Perhaps it could be good to put this parameter with elevated value as a recommended to the "Recommended OS Parameters page"

One other option to propose is to create a diagnostic tool which could check all kernel parameters and user limits  and give a report on which parameters could be adjusted for GP to perform best.

@interma  thank you so much for quickly jump in and directing me where to look for solution. Currently we understand this is definitely not a GP bug, but OS settings issue.
